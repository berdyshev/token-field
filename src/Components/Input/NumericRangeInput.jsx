import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import { Input } from 'deskpro-components/lib/Components/Forms';
import styles from '../../styles/style.css';
import TokenInput from './TokenInput';

export default class NumericRangeInput extends React.Component {
  static propTypes = {
    token: PropTypes.shape({
      type:  PropTypes.string,
      value: PropTypes.array.isRequired,
    }).isRequired,
    className:        PropTypes.string,
    unitPhrase:       PropTypes.string,
    convertToValue:   PropTypes.func,
    convertFromValue: PropTypes.func,
  };
  static defaultProps = {
    className:  '',
    unitPhrase: '',
    convertToValue(v) { return v; },
    convertFromValue(v) { return v; },
  };

  static moveCaretAtEnd(e) {
    const tempValue = e.target.value;
    e.target.value = '';
    e.target.value = tempValue;
  }

  constructor(props) {
    super(props);
    let from = '';
    let to = '';
    if (this.props.token.value.length) {
      from = this.props.token.value[0];
    }
    if (this.props.token.value.length > 1) {
      to = this.props.token.value[1];
    }
    this.state = {
      from,
      to,
    };
    this.handleChange = this.handleChange.bind(this);
    this.handleKeyDown = this.handleKeyDown.bind(this);
    this.getInput = this.getInput.bind(this);
    this.getValue = this.getValue.bind(this);
    this.focus = this.focus.bind(this);
  }

  getInput() {
    const { unitPhrase } = this.props;
    const { from, to } = this.state;
    return (
      <span>
        <Input
          value={this.props.convertFromValue(from)}
          name="from"
          className={classNames(styles.input)}
          onChange={this.handleChange}
          onKeyDown={this.handleKeyDown}
          ref={(c) => { this.inputFrom = c; }}
          onFocus={NumericRangeInput.moveCaretAtEnd}
          placeholder="__"
        />
        {unitPhrase} to
        <Input
          value={this.props.convertFromValue(to)}
          name="to"
          className={classNames(styles.input)}
          onChange={this.handleChange}
          onKeyDown={this.handleKeyDown}
          ref={(c) => { this.inputTo = c; }}
          onFocus={NumericRangeInput.moveCaretAtEnd}
          placeholder="__"
        />
        {unitPhrase}
      </span>
    );
  }

  getValue() {
    const { unitPhrase } = this.props;
    const { from, to } = this.state;
    let displayFrom;
    let displayTo;
    if (from) {
      displayFrom = this.props.convertFromValue(from);
    } else {
      displayFrom = '__';
    }
    if (to) {
      displayTo = this.props.convertFromValue(to);
    } else {
      displayTo = '__';
    }
    return `${displayFrom} ${unitPhrase} to ${displayTo} ${unitPhrase}`;
  }

  focus() {
    this.tokenInput.focus();
  }

  handleChange(value, name) {
    this.setState({
      [name]: this.props.convertToValue(value)
    });
  }

  handleKeyDown(e) {
    switch (e.key) {
      case 'Escape':
        this.setState({
          value: this.props.token.value
        });
        this.tokenInput.disableEditMode();
        break;
      case 'Tab':
      case 'Enter':
        if (!e.shiftKey) {
          if (e.target.name === 'from') {
            this.inputTo.focus();
          } else {
            this.tokenInput.disableEditMode();
          }
        } else if (e.target.name === 'to') {
          this.inputFrom.focus();
        } else {
          this.tokenInput.disableEditMode();
        }
        break;
      default:
        // Do nothing
        return;
    }
    e.preventDefault();
  }

  render() {
    const { token, className } = this.props;
    return (
      <TokenInput
        ref={(c) => { this.tokenInput = c; }}
        className={className}
        type={token.type}
        getInput={this.getInput}
        getValue={this.getValue}
      />
    );
  }
}
