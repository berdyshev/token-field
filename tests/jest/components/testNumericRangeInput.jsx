import React from 'react';
import { shallow, mount } from 'enzyme';
import { expect } from 'chai';
import NumericRangeInput from 'Components/Input/NumericRangeInput';

let wrapper;
let token;


const convertFromValue = value => Math.round(value / 1024 / 1024);
const convertToValue = value => value * 1024 * 1024;

beforeEach(() => {
  token = {
    type:  'user-message',
    value: [12582912, 20971520]
  };
  wrapper = mount(
    <NumericRangeInput
      token={token}
      unitPhrase="MB"
      className="test"
      convertToValue={convertToValue}
      convertFromValue={convertFromValue}
    />
  );
});

it('renders without crashing', () => {
  const textInput = shallow(<NumericRangeInput token={token} />);
  if (!expect(textInput).exist) {
    return false;
  }
  return undefined;
});

it('should render the input label', () => {
  expect(wrapper.contains(<div className="dp-code label">{token.type}:</div>)).to.equal(true);
});

it('should render the input value', () => {
  const span = wrapper.find('span');
  expect(span.text()).to.equal('12 MB to 20 MB');
});

it('should render empty value', () => {
  const emptyToken = {
    type:  'user-message',
    value: []
  };
  const emptyWrapper = mount(
    <NumericRangeInput
      token={emptyToken}
      unitPhrase="MB"
      className="test"
      convertToValue={convertToValue}
      convertFromValue={convertFromValue}
    />
  );
  const span = emptyWrapper.find('span');
  expect(span.text()).to.equal('__ MB to __ MB');
});

it('should have the class passed', () => {
  expect(wrapper.first().hasClass('test')).to.equal(true);
});
