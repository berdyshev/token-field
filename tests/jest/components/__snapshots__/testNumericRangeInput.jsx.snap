// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`+++capturing Snapshot of NumericRangeInput 1`] = `
<div
  className="token test"
>
  <div
    className="dp-code label"
  >
    user-message
    :
  </div>
  <span
    className="value"
    onClick={[Function]}
  >
    12582912 MB to 20971520 MB
  </span>
  <div
    className="token-remove remove"
    onClick={[Function]}
  >
    X
  </div>
</div>
`;

exports[`+++capturing Snapshot of NumericRangeInput with custom convert 1`] = `
<div
  className="token test"
>
  <div
    className="dp-code label"
  >
    user-message
    :
  </div>
  <span
    className="value"
    onClick={[Function]}
  >
    12 MB to 20 MB
  </span>
  <div
    className="token-remove remove"
    onClick={[Function]}
  >
    X
  </div>
</div>
`;

exports[`+++capturing Snapshot of NumericRangeInput with empty value 1`] = `
<div
  className="token test"
>
  <div
    className="dp-code label"
  >
    user-message
    :
  </div>
  <span
    className="value"
    onClick={[Function]}
  >
    __ MB to __ MB
  </span>
  <div
    className="token-remove remove"
    onClick={[Function]}
  >
    X
  </div>
</div>
`;
